### DISCUSSION TITLE
Species distribution data columns

### DISCUSSION THREAD

**Discussion by BioNCphilic** [earthlab-education/Earth-Analytics-AY24#520] [Q&A]
I'm falling behind and having trouble with data columns. In the species-distribution-32-normalize data notebook. The column 'index_right' line 11 does not work the way it does in the textbook video. The error message is KeyError: "['index_right'] not in index"

![image](https://github.com/user-attachments/assets/bbcaa3e6-a1f1-4bde-9642-221b25591841)

**Comment by gpb3037** [comment]
I had problems too but went back to the video in https://cu-esiil-edu.github.io/esiil-learning-portal/foundations/notebooks/03-species-distribution/species-distribution.html and in the example this now works (after some other iterations)
NOTE some of my variable names may be different than yours...

gbif_ecoregion_gdf = (
    eco_gdf
    # Match the CRS of the GBIF data and the ecoregions
    .to_crs(species_gbif_gdf.crs)
    # Find ecoregion for each observation
    .sjoin(
        species_gbif_gdf,
        how='inner', 
        predicate='contains')
    # Select the required columns
        [['gbifID','ECO_NAME','month']]
        .reset_index()
        .rename(columns={
            'index': 'ecoregion',
            'gbifID': 'observation_ID'})
)
gbif_ecoregion_gdf

**Reply by gpb3037** [reply]
shows up as this since I wanted to see the ecoregion name
	ecoregion	observation_ID	ECO_NAME	month
0	12	4743927038	Alberta-British Columbia foothills forests	5
1	12	4621947377	Alberta-British Columbia foothills forests	5
2	12	4761090115	Alberta-British Columbia foothills forests	6
3	12	4765238615	Alberta-British Columbia foothills forests	7
4	12	4630693711	Alberta-British Columbia foothills forests	6
...	...	...	...	...
159532	839	4633848077	North Atlantic moist mixed forests	10
159533	839	4749131402	North Atlantic moist mixed forests	9
159534	839	4763942306	North Atlantic moist mixed forests	9
159535	839	4746476478	North Atlantic moist mixed forests	9
159536	839	4719267065	North Atlantic moist mixed forests	9

**Reply by BioNCphilic** [reply]
Yep - tried that and based upon the textbook vide, eco name has the same issue as some other columns in that the values are repeated and not unique to each entry.

**Comment by gpb3037** [comment]
But then my next section doesn't work with observation_ID derived from my gbifID so something also still wierd.

**Reply by gpb3037** [reply]
it was because I didnt have square brackets in the next section 
I had .groupby('ecoregion','month')
when I should have had
.groupby(['ecoregion', 'month'])
so now this code works to do the summaries with that above:

### chatgpt version - aha I was missing square brackets in syntax
### Grouping data by ecoregion and month, and counting occurrences
occurrence_df = (
    gbif_ecoregion_gdf
    # Group by both ecoregion and month
    .groupby(['ecoregion', 'month'])
    # Count the number of occurrences
    .agg(occurrences=('observation_id', 'count'))
    .reset_index()  # Resetting index to have a proper DataFrame
)

occurrence_df

**Reply by gpb3037** [reply]
all of this stuff may be in error for gbfID vs OBJECT_ID as I see later in the video!

**Reply by BioNCphilic** [reply]
OBJECTID has the same issue - there are no unique identifiers that I could find on the ecoregion_gdf dataset. I also tried ECO_NAME, ECO_BIOME_, ECO_ID, and finally NNH before throwing my hands up (instead of smashing the computer!). Hopefully class today will help. Thanks @gpb3037

**Comment by eculler** [comment]
@BioNCphilic @gpb3037 was this issue solved in office hours? If not let me know.

I am still not sure what caused this issue in the first place, since it seems to work the first time through and then fail when the same code is run again. But it seems like making sure that the column names do match with existing column names fixes it? `'index_right'` is something that gets created automatically when joining with something other than the index.

Also note that in the second video we go through how to fix this. I will correct the first video in a few weeks, but for now I want to make sure that the videos match -- so go ahead and go through all of them to fix this issue.

**Comment by BioNCphilic** [comment]
Thanks for the help with this. I think I've got it figured out.
